trie/utils_test.go

```diff
@@ -122,3 +123,123 @@ func TestTrieTracerNoop(t *testing.T) {
 		t.Fatalf("Unexpected deleted node tracked %d", len(trie.tracer.deleteList()))
 	}
 }
+func TestTrieTracePrevValue(t *testing.T) {
+	db := NewDatabase(rawdb.NewMemoryDatabase())
+	trie := NewEmpty(db)
+	trie.tracer = newTracer()
+
+	paths, blobs := trie.tracer.prevList()
+	if len(paths) != 0 || len(blobs) != 0 {
+		t.Fatalf("Nothing should be tracked")
+	}
+	// Insert a batch of entries, all the nodes should be marked as inserted
+	vals := []struct{ k, v string }{
+		{"do", "verb"},
+		{"ether", "wookiedoo"},
+		{"horse", "stallion"},
+		{"shaman", "horse"},
+		{"doge", "coin"},
+		{"dog", "puppy"},
+		{"somethingveryoddindeedthis is", "myothernodedata"},
+	}
+	for _, val := range vals {
+		trie.Update([]byte(val.k), []byte(val.v))
+	}
+	paths, blobs = trie.tracer.prevList()
+	if len(paths) != 0 || len(blobs) != 0 {
+		t.Fatalf("Nothing should be tracked")
+	}
+
+	// Commit the changes and re-create with new root
+	root, nodes, _ := trie.Commit(false)
+	if err := db.Update(NewWithNodeSet(nodes)); err != nil {
+		t.Fatal(err)
+	}
+	trie, _ = New(TrieID(root), db)
+	trie.tracer = newTracer()
+	trie.resolveAndTrack(root.Bytes(), nil)
+
+	// Load all nodes in trie
+	for _, val := range vals {
+		trie.TryGet([]byte(val.k))
+	}
+
+	// Ensure all nodes are tracked by tracer with correct prev-values
+	iter := trie.NodeIterator(nil)
+	seen := make(map[string][]byte)
+	for iter.Next(true) {
+		// Embedded nodes are ignored since they are not present in
+		// database.
+		if iter.Hash() == (common.Hash{}) {
+			continue
+		}
+		blob, err := trie.reader.nodeBlob(iter.Path(), iter.Hash())
```

Comment:
`seen[string(iter.Path())] = common.CopyBytes(iter.NodeBlob()) I saw they use iter.NodeBlob()` here ? Do we implement it ?

Reply:
Source: https://github.com/ethereum/go-ethereum/commit/55430b6ea2bcc63bc11e7089690d90d8ac70dc5c
We haven't implemented this yet. They said that it is needed for path-based, but I'm not sure. Until now we can still get the node blob like this.

Reply:
Okay it is needed. I'll add it later.